platform/socket.hpp

```diff
@@ -1,31 +1,35 @@
 #pragma once
 
+#include "std/mutex.hpp"
 #include "std/string.hpp"
 #include "std/target_os.hpp"
-
-#if defined(OMIM_OS_IPHONE) || defined(OMIM_OS_MAC)
-@class SocketImpl;
-#else
-class SocketImpl;
-#endif
+#include "std/vector.hpp"
 
 namespace platform
 {
 class Socket
 {
 public:
-  Socket();
-  ~Socket();
-
-  bool Open(string const & host, uint16_t port);
-  void Close();
+  virtual ~Socket() {}
+  virtual bool Open(string const & host, uint16_t port) = 0;
+  virtual void Close() = 0;
 
-  bool Read(uint8_t * data, uint32_t count);
-  bool Write(uint8_t const * data, uint32_t count);
+  virtual bool Read(uint8_t * data, uint32_t count) = 0;
+  virtual bool Write(uint8_t const * data, uint32_t count) = 0;
 
-  void SetTimeout(uint32_t milliseconds);
+  virtual void SetTimeout(uint32_t milliseconds) = 0;
+};
 
-private:
-  SocketImpl * m_socketImpl = nullptr;
+class PlatformSocket final : public Socket
```

Comment:
I'm afraid that you need different subclasses on different platforms. I.e. you need SocketIOS, SocketAndroid, etc, because platform-specific sockets may have platform-specific fields.